cognite/client/api/assets.py

```diff
@@ -411,3 +454,146 @@ def search(
+    def __init__(self, assets: List[Asset], client: AssetsAPI):
+        self.validate_asset_hierarchy(assets)
+
+        self.remaining_assets = [Asset._load(a.dump(camel_case=True)) for a in assets]
+        self.client = client
+
+        self.number_of_assets = len(assets)
+
+        self.ref_id_to_remaining_children = self.initialize_ref_id_to_remaining_children_map(self.remaining_assets)
+        self.ref_id_to_id = {}
+
+        self.request_queue = queue.Queue()
+        self.response_queue = queue.Queue()
+
+        self.created_assets = AssetList([])
+
+    @staticmethod
+    def validate_asset_hierarchy(assets: List[Asset]) -> None:
+        ref_ids = [asset.ref_id for asset in assets if asset.ref_id is not None]
+        ref_ids_set = set(ref_ids)
+        assert sorted(ref_ids) == sorted(list(ref_ids_set)), "Duplicate ref_ids found"
+        for asset in assets:
+            parent_ref = asset.parent_ref_id
+            if parent_ref:
+                assert parent_ref in ref_ids_set, "parent_ref_id '{}' does not point to anything".format(parent_ref)
+                assert asset.parent_id is None, "An asset has both parent_id and parent_ref_id set."
+
+    @staticmethod
+    def initialize_ref_id_to_remaining_children_map(assets: List[Asset]) -> Dict[str, List[Asset]]:
+        ref_id_to_children = {asset.ref_id: [] for asset in assets if asset.ref_id is not None}
+        for asset in assets:
+            if asset.parent_ref_id in ref_id_to_children:
+                ref_id_to_children[asset.parent_ref_id].append(asset)
+        return ref_id_to_children
+
+    def get_descendants_unblocked_by_ref_id(self, assets: List[Asset], limit):
+        unblocked_children = []
+        for asset in assets:
+            if asset.ref_id in self.ref_id_to_remaining_children:
+                asset_children = self.ref_id_to_remaining_children[asset.ref_id].copy()
+                for child_asset in asset_children:
+                    unblocked_children.append(child_asset)
+                    self.ref_id_to_remaining_children[asset.ref_id].remove(child_asset)
+                    if len(unblocked_children) == limit:
+                        return unblocked_children
+
+                    max_descendents = limit - len(unblocked_children)
+                    unblocked_children.extend(self.get_descendants_unblocked_by_ref_id([child_asset], max_descendents))
+                    if len(unblocked_children) == limit:
+                        return unblocked_children
+        return unblocked_children
+
+    def get_unblocked_assets(self, limit) -> List[List[Asset]]:
+        unblocked_assets_lists = []
+        unblocked_assets = []
+
+        for asset in self.remaining_assets:
+            if asset.parent_ref_id is None or asset.parent_id is not None:
+                unblocked_assets.append(asset)
+            elif self.ref_id_to_id.get(asset.parent_ref_id) is not None:
+                asset.parent_id = self.ref_id_to_id[asset.parent_ref_id]
```

Comment:
Does this violate the constraint of not having `parent_id` and `parent_ref_id` at the same time?